contracts/ronin/validator/CoinbaseExecution.sol

```diff
@@ -286,25 +300,28 @@ abstract contract CoinbaseExecution is
    * Emits the `StakingRewardDistributionFailed` once the contract fails to distribute reward.
    *
    * Note: This method should be called once in the end of each period.
-   *
+   * - `delegatingFFRewards` is the fast finality rewards for delegators.
+   * - `delegatingRewards` already includes the fast finality rewards for delegators.
    */
   function _settleAndTransferDelegatingRewards(
     uint256 period,
     address[] memory currValidatorIds,
     uint256 totalDelegatingReward,
-    uint256[] memory delegatingRewards
+    uint256[] memory delegatingRewards,
+    uint256[] memory delegatingFFRewards
   ) private {
     IStaking _staking = IStaking(getContract(ContractType.STAKING));
     if (totalDelegatingReward > 0) {
       if (_unsafeSendRON(payable(address(_staking)), totalDelegatingReward)) {
         _staking.execRecordRewards(currValidatorIds, delegatingRewards, period);
+        emit FastFinalityRewardDistributed(currValidatorIds, delegatingFFRewards);
         emit StakingRewardDistributed(totalDelegatingReward, currValidatorIds, delegatingRewards);
         return;
       }
 
-      emit StakingRewardDistributionFailed(
-        totalDelegatingReward, currValidatorIds, delegatingRewards, address(this).balance
-      );
+      uint256 selfBalance = address(this).balance;
```

Comment:
I don't think cache this.balance consume less gas than call it directly.

Reply:
we actually save 12 gas.
Snippet to verify:
```solidity
contract GasSelfBalance {
    event Event(uint256 selfBalance);

    receive() external payable {}

    function noCache() public {
        emit Event(address(this).balance);
        emit Event(address(this).balance);
    }

    function cache() public {
        uint256 selfBalance = address(this).balance;

        emit Event(selfBalance);
        emit Event(selfBalance);
    }
}
```

Reply:
excellent